### PR TITLE
[ONNX] Cast Gather index to Long if needed

### PR DESCRIPTION
**Pull request by jiafatom**
Onnx op Gather index need be int32 or int64. However, we don't have this Cast in our converter.
Therefore, it fails the following UT (for opset 11+)
`seq_length.type().scalarType()` is None, so `_arange_cast_helper()` cannot treat it as all integral, then it will cast all to float. Then this float value will be used as Gather index, hence it throws error in ORT about float type index.
The fix is that we need cast Gather index type to Long if it is not int/long.

### CODE REVIEW

File: torch/onnx/symbolic_opset11.py [hunk 1]
```diff
@@ -57,6 +57,9 @@ def clamp_max(g, self, max):
 # Opset 11 gather accepts negative indices
 @parse_args('v', 'i', 'v')
 def select(g, self, dim, index):
+    index_scalar_type = index.type().scalarType()
```
Comment (KsenijaS) [inline torch/onnx/symbolic_opset11.py:60]: if shape_inference is turned on, will index_scalar_type still be None?

**Comment by KsenijaS** [comment]
looks good! LGTM

**Review by neginraoof — APPROVED** [review]
Overall looks good. 
Just curious, is the use of loop in the unit test necessary?
Also, we can create a helper to check if tensor is integral (similar to _is_fp) if we see more uses.

**Comment by jiafatom** [comment]
> Overall looks good.
> Just curious, is the use of loop in the unit test necessary?
> Also, we can create a helper to check if tensor is integral (similar to _is_fp) if we see more uses.

The loop corresponds to arange conversion, and in opset 11, there is logic in `_arange_cast_helper()`, In this UT,
`end.type().scalarType()=None`, so `_arange_cast_helper()` cannot treat it as all integral, then it will cast all to float. Since this is further used in Gather input index, the original Gather fails in ORT because index is casted as float.

**Review by bzinodev — APPROVED** [review]

**Comment by mruberry** [comment]
Unlanding. This broke pytorch_linux_xenial_py3_clang7_onnx_ort_test2. Relevant snippet:

```
=================================== FAILURES ===================================
Dec 01 16:34:30 _______________ TestONNXRuntime_opset11.test_gather_float_index ________________
Dec 01 16:34:30 
Dec 01 16:34:30 self = <test_pytorch_onnx_onnxruntime.TestONNXRuntime_opset11 testMethod=test_gather_float_index>
Dec 01 16:34:30 
Dec 01 16:34:30     def wrapper(self):
Dec 01 16:34:30         if self.opset_version < min_opset_version:
Dec 01 16:34:30             raise unittest.SkipTest("Skip verify test for unsupported opset_version")
Dec 01 16:34:30 >       return func(self)
Dec 01 16:34:30 
Dec 01 16:34:30 test/onnx/test_pytorch_common.py:49: 
Dec 01 16:34:30 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Dec 01 16:34:30 test/onnx/test_pytorch_onnx_onnxruntime.py:2366: in test_gather_float_index
Dec 01 16:34:30     self.run_test(model, (x, y))
Dec 01 16:34:30 test/onnx/test_pytorch_onnx_onnxruntime.py:150: in run_test
Dec 01 16:34:30     _run_test(model)
Dec 01 16:34:30 test/onnx/test_pytorch_onnx_onnxruntime.py:145: in _run_test
Dec 01 16:34:30     fixed_batch_size=fixed_batch_size)
Dec 01 16:34:30 test/onnx/test_pytorch_onnx_onnxruntime.py:104: in run_model_test
Dec 01 16:34:30     ort_compare_with_pytorch(ort_outs, output, rtol, atol)
Dec 01 16:34:30 test/onnx/test_pytorch_onnx_onnxruntime.py:71: in ort_compare_with_pytorch
Dec 01 16:34:30     [np.testing.assert_allclose(out, ort_out, rtol=rtol, atol=atol) for out, ort_out in zip(outputs, ort_outs)]
Dec 01 16:34:30 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Dec 01 16:34:30 
Dec 01 16:34:30 .0 = <zip object at 0x7f14f6c89bc8>
Dec 01 16:34:30 
Dec 01 16:34:30 >   [np.testing.assert_allclose(out, ort_out, rtol=rtol, atol=atol) for out, ort_out in zip(outputs, ort_outs)]
Dec 01 16:34:30 E   AssertionError: 
Dec 01 16:34:30 E   Not equal to tolerance rtol=0.001, atol=1e-07
Dec 01 16:34:30 E   
Dec 01 16:34:30 E   Mismatched elements: 13 / 17 (76.5%)
Dec 01 16:34:30 E   Max absolute difference: 4
Dec 01 16:34:30 E   Max relative difference: 3.
Dec 01 16:34:30 E    x: array(4)
Dec 01 16:34:30 E    y: array([4, 0, 3, 3, 2, 3, 0, 4, 4, 3, 1, 3, 1, 2, 3, 4, 1], dtype=int64)
Dec 01 16:34:30 
Dec 01 16:34:30 test/onnx/test_pytorch_onnx_onnxruntime.py:71: AssertionError
Dec 01 16:34:30 ----------------------------- Captured stderr call -----------------------------
Dec 01 16:34:30 2020-12-01 16:27:30.506515944 [W:onnxruntime:, graph.cc:1031 Graph] Initializer 31 appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
Dec 01 16:34:30 2020-12-01 16:27:30.506584011 [W:onnxruntime:, graph.cc:1031 Graph] Initializer 32 appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
```

**Comment by jiafatom** [comment]
@mruberry Thanks for pointing this out. It seems that this PR passed all CI build in the beginning (the previous last commit was tested 29 days ago). Within these 29 days, new code are merged, and there seems no issue before merge. Now we see something broken, so you unland it.
I just made a fix which works on my local machine. So I update the commit. 
Would you like to merge it once all the checks go green? Thank you!

**Comment by mruberry** [comment]
> @mruberry Thanks for pointing this out. It seems that this PR passed all CI build in the beginning (the previous last commit was tested 29 days ago). Within these 29 days, new code are merged, and there seems no issue before merge. Now we see something broken (based on latest master), so you unland it.

I think your analysis is correct. Sorry, unlanding is always a pain, but it needs to be so the CI continues to work properly.

> I just made a fix which works on my local machine. Since the original code is in master, so I update my fix in this PR.
> Would you like to merge it once all the checks go green? Thank you!

@bzinodev will help you get this PR landed again.

**Comment by jiafatom** [comment]
@mruberry CI build looks good to me. (only codecov/patch is failing but I see this failing in other PRs too. All other 87 checks are successful.) Can we escalate merge it? The original commit is still in master, and it is failing for other PRs.

According to @spandantiwari , @bzinodev will work on other tasks. We need someone to help merge this PR. Thanks!

**Comment by mruberry** [comment]
> According to @spandantiwari , @bzinodev will work on other tasks. We need someone to help merge this PR. Thanks!

Sure. I can help you.

> Another thing I heard is that you usually revert the original PR and I open a new PR. Currently I just add commits to this original PR because it is reopened and not reverted. But either way works for me.

Both a new PR or updating this one is fine. Let's land this one since it's updated.

**Comment by spandantiwari** [comment]
Thank you @mruberry .

**Pull request closed without merging**